File: Stakeholders/app/services/consumer.py
import asyncio
import json
import logging
import os

import aio_pika
from aio_pika import Message

from app.schemas.profile import BalanceResponse
from app.services.profile import ProfileService

QUEUE_STAKEHOLDER_COMMANDS = "saga.stakeholder.commands"
INTERNAL_SECRET = os.getenv("INTERNAL_SECRET")

logger = logging.getLogger(__name__)


class StakeholderConsumer:
    def __init__(self, profile_service: ProfileService):
        self.profile_service = profile_service
        self.connection = None
        self.channel = None

    async def connect(self, amqp_url: str):
        self.connection = await aio_pika.connect_robust(amqp_url)
        self.channel = await self.connection.channel()
        await self.channel.set_qos(prefetch_count=10)
        queue = await self.channel.declare_queue(QUEUE_STAKEHOLDER_COMMANDS, durable=True)
        await queue.consume(self.handle)
        logger.info(f"[MQ] Stakeholder consumer started on {QUEUE_STAKEHOLDER_COMMANDS}")

    async def handle(self, message: aio_pika.IncomingMessage):
        async with message.process():
            try:
                cmd = json.loads(message.body)
                command = cmd.get("command")
                payload = cmd.get("payload", {})
                is_internal = cmd.get("is_internal", False)

                match command:
                    case "DEDUCT_BALANCE":
                        reply = await self._handle_deduct_balance(payload)
                    case "REFUND_BALANCE":
                        reply = await self._handle_refund_balance(payload, is_internal)
                    case _:
                        reply = {"success": False, "error": f"unknown command: {command}"}

                await self.channel.default_exchange.publish(
                    Message(
                        body=json.dumps(reply).encode(),
                        content_type="application/json",
                        correlation_id=message.correlation_id,
                    ),
                    routing_key=message.reply_to,
                )
            except Exception as e:
                logger.error(f"[MQ] Unhandled error in consumer: {e}")

    async def _handle_deduct_balance(self, payload: dict) -> dict:
        try:
            user_id = int(payload["tourist_id"])
            amount = float(payload["amount"])
            result = await self.profile_service.deduct_balance(user_id, amount)
            return {"success": True, "payload": {"new_balance": result.balance}}
        except Exception as e:
            return {"success": False, "error": str(e)}

    async def _handle_refund_balance(self, payload: dict, is_internal: bool) -> dict:
        if not is_internal:
            return {"success": False, "error": "unauthorized: internal only"}
        try:
            user_id = int(payload["tourist_id"])
            amount = float(payload["amount"])
            result = await self.profile_service.refund_balance(user_id, amount)
            return {"success": True, "payload": {"new_balance": result.balance}}
        except Exception as e:
            return {"success": False, "error": str(e)}

    async def close(self):
        if self.connection:
            await self.connection.close()

File: Authorization/consumer.go
package main

import (
	"context"
	"encoding/json"
	"log"

	amqp "github.com/rabbitmq/amqp091-go"
)

const QueueAuthCommands = "saga.auth.commands"

type SagaCommand struct {
	Command    string          `json:"command"`
	IsInternal bool            `json:"is_internal"`
	Payload    json.RawMessage `json:"payload"`
}

type SagaReply struct {
	Success bool            `json:"success"`
	Error   string          `json:"error,omitempty"`
	Payload json.RawMessage `json:"payload,omitempty"`
}

type AuthConsumer struct {
	conn *amqp.Connection
	ch   *amqp.Channel
	srv  *Server
}

func NewAuthConsumer(amqpURL string, srv *Server) (*AuthConsumer, error) {
	conn, err := amqp.Dial(amqpURL)
	if err != nil {
		return nil, err
	}
	ch, err := conn.Channel()
	if err != nil {
		return nil, err
	}
	if _, err := ch.QueueDeclare(QueueAuthCommands, true, false, false, false, nil); err != nil {
		return nil, err
	}
	ch.Qos(1, 0, false)
	return &AuthConsumer{conn: conn, ch: ch, srv: srv}, nil
}

func (c *AuthConsumer) Start() {
	msgs, _ := c.ch.Consume(QueueAuthCommands, "", false, false, false, false, nil)
	log.Println("[MQ] Auth consumer started")
	for d := range msgs {
		go c.handle(d)
	}
}

func (c *AuthConsumer) Close() {
	c.ch.Close()
	c.conn.Close()
}

func (c *AuthConsumer) handle(d amqp.Delivery) {
	var cmd SagaCommand
	if err := json.Unmarshal(d.Body, &cmd); err != nil {
		c.reply(d, SagaReply{Success: false, Error: "invalid command payload"})
		d.Nack(false, false)
		return
	}

	var reply SagaReply
	switch cmd.Command {
	case "REGISTER_USER":
		reply = c.handleRegister(cmd)
	case "DELETE_USER":
		reply = c.handleDelete(cmd)
	default:
		reply = SagaReply{Success: false, Error: "unknown command: " + cmd.Command}
	}

	c.reply(d, reply)
	d.Ack(false)
}

func (c *AuthConsumer) handleRegister(cmd SagaCommand) SagaReply {
	var req RegisterRequest
	if err := json.Unmarshal(cmd.Payload, &req); err != nil {
		return SagaReply{Success: false, Error: "invalid payload"}
	}
	result, err := c.srv.registerUser(context.Background(), req)
	if err != nil {
		return SagaReply{Success: false, Error: err.Error()}
	}
	payload, _ := json.Marshal(result)
	return SagaReply{Success: true, Payload: payload}
}

func (c *AuthConsumer) handleDelete(cmd SagaCommand) SagaReply {
	if !cmd.IsInternal {
		return SagaReply{Success: false, Error: "unauthorized: internal only"}
	}
	var req struct {
		UserID int64 `json:"user_id"`
	}
	if err := json.Unmarshal(cmd.Payload, &req); err != nil {
		return SagaReply{Success: false, Error: "invalid payload"}
	}
	if err := c.srv.deleteUser(context.Background(), req.UserID); err != nil {
		return SagaReply{Success: false, Error: err.Error()}
	}
	return SagaReply{Success: true}
}

func (c *AuthConsumer) reply(d amqp.Delivery, reply SagaReply) {
	body, _ := json.Marshal(reply)
	c.ch.Publish("", d.ReplyTo, false, false, amqp.Publishing{
		ContentType:   "application/json",
		CorrelationId: d.CorrelationId,
		Body:          body,
	})
}
